### ISSUE TITLE
New Generator: Swift5 client generator with multiple HTTP library support 

### ISSUE THREAD

**Issue by wing328** [OpenAPITools/openapi-generator#2721]
##### Description

We would like to start a discussion on a new Swift5 client generator that supports multiple HTTP libraries.

Currently, the Swift (2.x, 3.x, 4x) generators use [Swift Alamofire](https://github.com/Alamofire/Alamofire). For Swift 5, we want to add a new generator that can supports different HTTP libraries via the `library` option similar to what we've done in the Java client generator.

The implementation can be a swappable API client class (the approach we used in Java) or other approach that makes sense in Swift.

This is not to say we'll no longer use Swift Alamofire, which is still a decent choice, so we may want to add a new swift5-generator to support Swift 5.x (only) for time being while we work on the new generator with multiple HTTP library support.

cc @d-date

**Comment by markfinlabs** [comment]
A great idea. There is obviously URLSession in Foundation and now, for server-side Swift, seems a new HTTP client library is in development....
https://forums.swift.org/t/generic-http-client-library-pitch/23341
https://github.com/swift-server/swift-nio-http-client
Our interest would be in this new library for server-side Swift.

**Comment by mac2000** [comment]
While this is in progress is there any known workaround for swift4 generated code to be working in swift5? (e.g. by modifying templates and/or podspec)

**Comment by martinwiboe** [comment]
@mac2000 the swift4 generated code is working fine for me using swift 5. What error are you seeing?

**Comment by d-date** [comment]
Don't worry. Swift4 generator still can be used for existing users (including need objcCompatible option). Swift 5 and Swift 5.1 has so many great features for OAS 3.0. Until Swift 4 generator we have only for OAS 2.0(Swagger).

For various API Clients, this generator only provide interface and protocols with `onlyPalinInterface` option default is `true` . So we can use URLSession in Foundation, Alamofire and any other API Clients.

**Comment by d-date** [comment]
I'd like to ask you how many PromiseKit users exist now? From Swift 5, Result is introduced in Standard Library. If you have no problems, I want to deprecate PromiseKit support since this generator has too many option to keep maintain.

**Comment by markfinlabs** [comment]
We're not using PromiseKit and also are trying to adopt the latest Swift features 😀

**Comment by lightbow** [comment]
AlamoFire was fun while it lasted, but I'd love one fewer dependency. A generator using just URLSession / Foundation would be wonderful.

**Comment by spprichard** [comment]
Is there any progress on this? 

I have a requirement that the networking layer of the generated code does not use Alamofire. Due to Alamofire not being compatible with Linux. Being able to swap out the networking layer of the generated code for something that can run on Linux is what I would need for my current project. Is this possible?

**Comment by 4brunu** [comment]
@spprichard What library or so will you use on Linux to make the network requests?

The Swift 4 generator already allows you to use a custom network stack with the network library that you want, but it will take some work.

If you want to go that path, you will need to subclass [RequestBuilder]( https://github.com/OpenAPITools/openapi-generator/blob/8cf34ca80f380a1ae5cee9e4360a1c21762e7ec6/samples/client/petstore/swift4/default/PetstoreClient/Classes/OpenAPIs/APIs.swift#L16) to implement the network stack with the library that you want, create an implementation for [RequestBuilderFactory](https://github.com/OpenAPITools/openapi-generator/blob/8cf34ca80f380a1ae5cee9e4360a1c21762e7ec6/samples/client/petstore/swift4/default/PetstoreClient/Classes/OpenAPIs/APIs.swift#L58) that returns your new network stack implementation, and then swap the network factory in [PetstoreClientAPI.requestBuilderFactory](https://github.com/OpenAPITools/openapi-generator/blob/8cf34ca80f380a1ae5cee9e4360a1c21762e7ec6/samples/client/petstore/swift4/default/PetstoreClient/Classes/OpenAPIs/APIs.swift#L13), but again it will take some work and the Alamofire will still be there, but it won't do anything.

Hope that I didn't say something wrong.

**Comment by spprichard** [comment]
Hey @4brunu,

Any network framework that can compile for Linux. Is there anyway to exclude Alamofire altogether? Also, is there a guide for this?

**Comment by 4brunu** [comment]
You can do this manually, but it's not ideal.
The other way is using library option to choose the http stack.
The kotlin generator, uses the library option to choose the http library for example.

**Comment by spprichard** [comment]
@4brunu Where can I find docs on which libraries are supported for swift4? I see that I can pass a value to the `--library` flag, but don't know what values it can take. Here is the command I am trying currently
`swagger-codegen generate -i http://petstore.swagger.io/v2/swagger.json -l swift4 --library PromiseKit -o /tmp/test/`

This was done after installing swagger-codegen via brew, as I am on MacOS

**Comment by 4brunu** [comment]
If anyone would like to help, I'm trying to implement support for URLSession.
I created this repo with an example of the generated code.
https://github.com/4brunu/openapi-generator-swift-nsurlsession

This is the class with the Alamofire implementation
https://github.com/4brunu/openapi-generator-swift-nsurlsession/blob/master/openapi/OpenAPIClient/Classes/OpenAPIs/AlamofireImplementations.swift

I'm trying to create an equivalent implementation that instead of Alamofire used URLSession.
This is the URLSession implementation that I started doing
https://github.com/4brunu/openapi-generator-swift-nsurlsession/blob/master/openapi/OpenAPIClient/Classes/OpenAPIs/URLSessionImplementations.swift

If anyone would like to help, any help is welcome.
🙂

**Comment by Christian-Seiler** [comment]
@4brunu I've also started to look into an implementation here: [Christian-Seiler/openapi-generator](https://github.com/Christian-Seiler/openapi-generator/tree/feature/swift-5)

I would love to help, however I'm not sure how much I could contribute.

**Comment by 4brunu** [comment]
I'm also not sure how to split this into tasks.
But since yesterday I made some progress with the URLSession implementation.
I would appreciate the code review and testing once it's done.

**Comment by 4brunu** [comment]
For those interested, I just created the PR https://github.com/OpenAPITools/openapi-generator/pull/4802 that add support for URLSession instead of Alamofire.
It's possible to choose between Alamofire and URLSession with the library option.
Since this is a big PR, I would appreciate any feedback and help with testing.

**Comment by Christian-Seiler** [comment]
I've tested your Code with my API, and have no issues so far.

One thing I'd like to note however. Could you add CaseIterable to the enum model template?

**Comment by 4brunu** [comment]
@Christian-Seiler Just to make sure, you did test the URLSession implementation, right? (I'm asking because I have several open PR's right now 🙂 )
I have also created https://github.com/OpenAPITools/openapi-generator/pull/4843 to add CaseIterable to the enum's.

**Comment by Christian-Seiler** [comment]
@4brunu Yes. I'm testing #4802 

I've run in an issue there:

When calling ```routeGetWithRequestBuilder().execute{ ... }```

```Requestbuilder.execute``` gets called instead of ```URLSessionRequestBuilder.execute```

If I separate the initialisation and execute into:

        let builder = routeGetWithRequestBuilder()

        print(builder)

        builder.execute { (response, error) -> Void in
            completion(response?.body, error)
        }

```builder``` has the correct type: ```URLSessionDecodableRequestBuilder<Response>```

**Comment by 4brunu** [comment]
Hum that's weird... 
I have tested it now and it's working.
It's works also on the units tests.
Check here for example

https://github.com/4brunu/openapi-generator/blob/cfc33d3c7314a8874d9fb90cd673226f828082ce/samples/client/petstore/swift4/urlsessionLibrary/SwaggerClientTests/SwaggerClientTests/PetAPITests.swift#L33-L40

Can you create a sample project that show's the issue please?

**Comment by Christian-Seiler** [comment]
> Hum that's weird...
> I have tested it now and it's working.
> It's works also on the units tests.
> Check here for example
> 
> https://github.com/4brunu/openapi-generator/blob/cfc33d3c7314a8874d9fb90cd673226f828082ce/samples/client/petstore/swift4/urlsessionLibrary/SwaggerClientTests/SwaggerClientTests/PetAPITests.swift#L33-L40
> 
> Can you create a sample project that show's the issue please?

I've got it to work.

The issue was when I added the api as a swift package to the project. Once I created an Xcode project, built the framework and imported that, the result came in correctly.

**Comment by 4brunu** [comment]
***TLDR:*** In the new Swift 5 generator, which http client should be the default option? URLSession or Alamofire?

Hey, I have created a Swift 5 generator here https://github.com/OpenAPITools/openapi-generator/pull/4086.
But I have one question that I would like to have your opinions.
In this new generator, I have introduced support for URLSession.
My question is, which one should be the default http client? URLSession or Alamofire?
I have tested URLSession, and I'm using it in my project, and there are no know bugs, but it's new, so it can contain bugs.
In the long run, I think URLSession should be the default option, since it comes with the OS and doesn't require any dependencies.
But now, I'm not sure if we should default to URLSession now, or wait some time, to allow some more people to test it, and leave the alamofire as the default one, since the implementation it's almost the same as Swift 4 generator.
Please let me know what you think 🙂

**Comment by Christian-Seiler** [comment]
I think URLSession should be default.

I've tested the URLSession implementation and use it in my project. I have found no issues so far.

**Comment by 4brunu** [comment]
@Christian-Seiler thanks for your opinion, I have changed the default http client to URLSession.
If someone could please help to review the Swift 5 generator PR https://github.com/OpenAPITools/openapi-generator/pull/4086 I would appreciate 🙂
